File: src/test/java/cz/vhromada/catalog/utils/MusicUtils.java
package cz.vhromada.catalog.utils;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.SoftAssertions.assertSoftly;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.persistence.EntityManager;

import cz.vhromada.catalog.domain.Song;
import cz.vhromada.catalog.entity.Music;

/**
 * A class represents utility class for music.
 *
 * @author Vladimir Hromada
 */
public final class MusicUtils {

    /**
     * ID
     */
    public static final int ID = 1;

    /**
     * Count of music
     */
    public static final int MUSIC_COUNT = 3;

    /**
     * Position
     */
    public static final int POSITION = 10;

    /**
     * Music name
     */
    private static final String MUSIC = "Music ";

    /**
     * Creates a new instance of MusicUtils.
     */
    private MusicUtils() {
    }

    /**
     * Returns music.
     *
     * @param id ID
     * @return music
     */
    public static cz.vhromada.catalog.domain.Music newMusicDomain(final Integer id) {
        final cz.vhromada.catalog.domain.Music music = new cz.vhromada.catalog.domain.Music();
        updateMusic(music);
        if (id != null) {
            music.setId(id);
            music.setPosition(id - 1);
        }
        music.setSongs(Collections.emptyList());

        return music;
    }

    /**
     * Returns music with songs.
     *
     * @param id ID
     * @return music with songs
     */
    public static cz.vhromada.catalog.domain.Music newMusicWithSongs(final Integer id) {
        final Song song = SongUtils.newSongDomain(id);
        if (id == null) {
            song.setPosition(0);
        }

        final cz.vhromada.catalog.domain.Music music = newMusicDomain(id);
        music.setSongs(Collections.singletonList(song));

        return music;
    }

    /**
     * Updates music fields.
     *
     * @param music music
     */
    @SuppressWarnings("Duplicates")
    public static void updateMusic(final cz.vhromada.catalog.domain.Music music) {
        music.setName("Name");
        music.setWikiEn("enWiki");
        music.setWikiCz("czWiki");
        music.setMediaCount(1);
        music.setNote("Note");
    }

    /**
     * Returns music.
     *
     * @param id ID
     * @return music
     */
    public static Music newMusic(final Integer id) {
        final Music music = new Music();
        updateMusic(music);
        if (id != null) {
            music.setId(id);
            music.setPosition(id - 1);
        }

        return music;
    }

    /**
     * Updates music fields.
     *
     * @param music music
     */
    @SuppressWarnings("Duplicates")
    public static void updateMusic(final Music music) {
        music.setName("Name");
        music.setWikiEn("enWiki");
        music.setWikiCz("czWiki");
        music.setMediaCount(1);
        music.setNote("Note");
    }

    /**
     * Returns music.
     *
     * @return music
     */
    public static List<cz.vhromada.catalog.domain.Music> getMusic() {
        final List<cz.vhromada.catalog.domain.Music> music = new ArrayList<>();
        for (int i = 0; i < MUSIC_COUNT; i++) {
            music.add(getMusic(i + 1));
        }

        return music;
    }

    /**
     * Returns music for index.
     *
     * @param index index
     * @return music for index
     */
    public static cz.vhromada.catalog.domain.Music getMusic(final int index) {
        final cz.vhromada.catalog.domain.Music music = new cz.vhromada.catalog.domain.Music();
        music.setId(index);
        music.setName(MUSIC + index + " name");
        music.setWikiEn(MUSIC + index + " English Wikipedia");
        music.setWikiCz(MUSIC + index + " Czech Wikipedia");
        music.setMediaCount(index * 10);
        music.setNote(index == 2 ? MUSIC + "2 note" : "");
        music.setPosition(index - 1);
        music.setSongs(SongUtils.getSongs(index));

        return music;
    }

    /**
     * Returns music.
     *
     * @param entityManager entity manager
     * @param id            music ID
     * @return music
     */
    public static cz.vhromada.catalog.domain.Music getMusic(final EntityManager entityManager, final int id) {
        return entityManager.find(cz.vhromada.catalog.domain.Music.class, id);
    }

    /**
     * Returns music with updated fields.
     *
     * @param id            music ID
     * @param entityManager entity manager
     * @return music with updated fields
     */
    public static cz.vhromada.catalog.domain.Music updateMusic(final EntityManager entityManager, final int id) {
        final cz.vhromada.catalog.domain.Music music = getMusic(entityManager, id);
        updateMusic(music);
        music.setPosition(POSITION);

        return music;
    }

    /**
     * Returns count of music.
     *
     * @param entityManager entity manager
     * @return count of music
     */
    public static int getMusicCount(final EntityManager entityManager) {
        return entityManager.createQuery("SELECT COUNT(m.id) FROM Music m", Long.class).getSingleResult().intValue();
    }

    /**
     * Asserts music deep equals.
     *
     * @param expected expected music
     * @param actual   actual music
     */
    public static void assertMusicDeepEquals(final List<cz.vhromada.catalog.domain.Music> expected, final List<cz.vhromada.catalog.domain.Music> actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertThat(expected.size()).isEqualTo(actual.size());
        if (!expected.isEmpty()) {
            for (int i = 0; i < expected.size(); i++) {
                assertMusicDeepEquals(expected.get(i), actual.get(i));
            }
        }
    }

    /**
     * Asserts music deep equals.
     *
     * @param expected expected music
     * @param actual   actual music
     */
    public static void assertMusicDeepEquals(final cz.vhromada.catalog.domain.Music expected, final cz.vhromada.catalog.domain.Music actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(expected.getId()).isEqualTo(actual.getId());
            softly.assertThat(expected.getName()).isEqualTo(actual.getName());
            softly.assertThat(expected.getWikiEn()).isEqualTo(actual.getWikiEn());
            softly.assertThat(expected.getWikiCz()).isEqualTo(actual.getWikiCz());
            softly.assertThat(expected.getMediaCount()).isEqualTo(actual.getMediaCount());
            softly.assertThat(expected.getNote()).isEqualTo(actual.getNote());
            softly.assertThat(expected.getPosition()).isEqualTo(actual.getPosition());
            if (expected.getSongs() == null) {
                softly.assertThat(actual.getSongs()).isNull();
            } else {
                SongUtils.assertSongsDeepEquals(expected.getSongs(), actual.getSongs());
            }
        });
    }

    /**
     * Asserts music deep equals.
     *
     * @param expected expected list of music
     * @param actual   actual music
     */
    public static void assertMusicListDeepEquals(final List<Music> expected, final List<cz.vhromada.catalog.domain.Music> actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertThat(expected.size()).isEqualTo(actual.size());
        if (!expected.isEmpty()) {
            for (int i = 0; i < expected.size(); i++) {
                assertMusicDeepEquals(expected.get(i), actual.get(i));
            }
        }
    }

    /**
     * Asserts music deep equals.
     *
     * @param expected expected music
     * @param actual   actual music
     */
    public static void assertMusicDeepEquals(final Music expected, final cz.vhromada.catalog.domain.Music actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isEqualTo(expected.getId());
            softly.assertThat(actual.getName()).isEqualTo(expected.getName());
            softly.assertThat(actual.getWikiEn()).isEqualTo(expected.getWikiEn());
            softly.assertThat(actual.getWikiCz()).isEqualTo(expected.getWikiCz());
            softly.assertThat(actual.getMediaCount()).isEqualTo(expected.getMediaCount());
            softly.assertThat(actual.getNote()).isEqualTo(expected.getNote());
            softly.assertThat(actual.getPosition()).isEqualTo(expected.getPosition());
        });
    }

}

File: src/test/kotlin/cz/vhromada/catalog/utils/MusicUtils.kt
package cz.vhromada.catalog.utils

import cz.vhromada.catalog.entity.Music
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly
import java.util.ArrayList
import javax.persistence.EntityManager

/**
 * Updates music fields.
 *
 * @return updated music
 */
fun cz.vhromada.catalog.domain.Music.updated(): cz.vhromada.catalog.domain.Music {
    return copy(name = "Name", wikiEn = "enWiki", wikiCz = "czWiki", mediaCount = 1, note = "Note", audit = AuditUtils.newAudit())
}

/**
 * Updates music fields.
 *
 * @return updated music
 */
fun Music.updated(): Music {
    return copy(name = "Name", wikiEn = "enWiki", wikiCz = "czWiki", mediaCount = 1, note = "Note")
}

/**
 * A class represents utility class for music.
 *
 * @author Vladimir Hromada
 */
object MusicUtils {

    /**
     * Count of music
     */
    const val MUSIC_COUNT = 3

    /**
     * Position
     */
    const val POSITION = 10

    /**
     * Music name
     */
    private const val MUSIC = "Music "

    /**
     * Returns music.
     *
     * @return music
     */
    fun getMusic(): List<cz.vhromada.catalog.domain.Music> {
        val music = ArrayList<cz.vhromada.catalog.domain.Music>()
        for (i in 0 until MUSIC_COUNT) {
            music.add(getMusic(i + 1))
        }

        return music
    }

    /**
     * Returns music.
     *
     * @param id ID
     * @return music
     */
    fun newMusicDomain(id: Int?): cz.vhromada.catalog.domain.Music {
        return cz.vhromada.catalog.domain.Music(
                id = id,
                name = "",
                wikiEn = null,
                wikiCz = null,
                mediaCount = 0,
                note = null,
                position = if (id == null) null else id - 1,
                songs = emptyList(),
                audit = null)
                .updated()
    }

    /**
     * Returns music with songs.
     *
     * @param id ID
     * @return music with songs
     */
    fun newMusicWithSongs(id: Int?): cz.vhromada.catalog.domain.Music {
        return newMusicDomain(id)
                .copy(songs = listOf(SongUtils.newSongDomain(id)))
    }

    /**
     * Returns music.
     *
     * @param id ID
     * @return music
     */
    fun newMusic(id: Int?): Music {
        return Music(
                id = id,
                name = "",
                wikiEn = null,
                wikiCz = null,
                mediaCount = 0,
                note = null,
                position = if (id == null) null else id - 1)
                .updated()
    }

    /**
     * Returns music for index.
     *
     * @param index index
     * @return music for index
     */
    fun getMusic(index: Int): cz.vhromada.catalog.domain.Music {
        return cz.vhromada.catalog.domain.Music(
                id = index,
                name = "$MUSIC$index name",
                wikiEn = "$MUSIC$index English Wikipedia",
                wikiCz = "$MUSIC$index Czech Wikipedia",
                mediaCount = index * 10,
                note = if (index == 2) MUSIC + "2 note" else "",
                position = index - 1,
                songs = SongUtils.getSongs(index),
                audit = AuditUtils.getAudit())
    }

    /**
     * Returns music.
     *
     * @param entityManager entity manager
     * @param id            music ID
     * @return music
     */
    fun getMusic(entityManager: EntityManager, id: Int): cz.vhromada.catalog.domain.Music? {
        return entityManager.find(cz.vhromada.catalog.domain.Music::class.java, id)
    }

    /**
     * Returns music with updated fields.
     *
     * @param id            music ID
     * @param entityManager entity manager
     * @return music with updated fields
     */
    fun updateMusic(entityManager: EntityManager, id: Int): cz.vhromada.catalog.domain.Music {
        return getMusic(entityManager, id)!!
                .updated()
                .copy(position = POSITION)
    }

    /**
     * Returns count of music.
     *
     * @param entityManager entity manager
     * @return count of music
     */
    @Suppress("CheckStyle")
    fun getMusicCount(entityManager: EntityManager): Int {
        return entityManager.createQuery("SELECT COUNT(m.id) FROM Music m", java.lang.Long::class.java).singleResult.toInt()
    }

    /**
     * Asserts music deep equals.
     *
     * @param expected expected music
     * @param actual   actual music
     */
    fun assertMusicDeepEquals(expected: List<cz.vhromada.catalog.domain.Music?>?, actual: List<cz.vhromada.catalog.domain.Music?>?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertThat(expected!!.size).isEqualTo(actual!!.size)
        if (expected.isNotEmpty()) {
            for (i in expected.indices) {
                assertMusicDeepEquals(expected[i], actual[i])
            }
        }
    }

    /**
     * Asserts music deep equals.
     *
     * @param expected expected music
     * @param actual   actual music
     */
    fun assertMusicDeepEquals(expected: cz.vhromada.catalog.domain.Music?, actual: cz.vhromada.catalog.domain.Music?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(expected!!.id).isEqualTo(actual!!.id)
            it.assertThat(expected.name).isEqualTo(actual.name)
            it.assertThat(expected.wikiEn).isEqualTo(actual.wikiEn)
            it.assertThat(expected.wikiCz).isEqualTo(actual.wikiCz)
            it.assertThat(expected.mediaCount).isEqualTo(actual.mediaCount)
            it.assertThat(expected.note).isEqualTo(actual.note)
            it.assertThat(expected.position).isEqualTo(actual.position)
            SongUtils.assertSongsDeepEquals(expected.songs, actual.songs)
            AuditUtils.assertAuditDeepEquals(expected.audit, actual.audit)
        }
    }

    /**
     * Asserts music deep equals.
     *
     * @param expected expected list of music
     * @param actual   actual music
     */
    fun assertMusicListDeepEquals(expected: List<Music?>?, actual: List<cz.vhromada.catalog.domain.Music?>?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertThat(expected!!.size).isEqualTo(actual!!.size)
        if (expected.isNotEmpty()) {
            for (i in expected.indices) {
                assertMusicDeepEquals(expected[i], actual[i])
            }
        }
    }

    /**
     * Asserts music deep equals.
     *
     * @param expected expected music
     * @param actual   actual music
     */
    fun assertMusicDeepEquals(expected: Music?, actual: cz.vhromada.catalog.domain.Music?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(actual!!.id).isEqualTo(expected!!.id)
            it.assertThat(actual.name).isEqualTo(expected.name)
            it.assertThat(actual.wikiEn).isEqualTo(expected.wikiEn)
            it.assertThat(actual.wikiCz).isEqualTo(expected.wikiCz)
            it.assertThat(actual.mediaCount).isEqualTo(expected.mediaCount)
            it.assertThat(actual.note).isEqualTo(expected.note)
            it.assertThat(actual.position).isEqualTo(expected.position)
        }
    }

}
